docs/design/cpsubsystem/04-cp-object-lifecycle.md

```diff
@@ -0,0 +1,319 @@
+- [CreateRaftObjectMessageTask](../../../hazelcast/src/main/java/com/hazelcast/cp/internal/datastructures/spi/client/CreateRaftObjectMessageTask.java)
+- [CreateOrGetRaftObjectOp](../../../hazelcast/src/main/java/com/hazelcast/cp/internal/datastructures/spi/operation/CreateOrGetRaftObjectOp.java)
+
+This flow creates an empty data structure of a specific type (serviceName) on the Server side with an UUID and returns this UUID to the Client side:
+
+```java
+private UUID getObjectId(RaftGroupId groupId, String serviceName, String objectName) {
+    ClientMessage request = CPGroupCreateCPObjectCodec.encodeRequest(groupId, serviceName, objectName);
+    ClientMessage response = new ClientInvocation(client, request, objectName).invoke().joinInternal();
+    return CPGroupCreateCPObjectCodec.decodeResponse(response);
+}
+```
+
+This UUID is recorded as a property of the object proxy on the Client side:
+
+```java
+public class AtomicLongProxy extends ClientProxy implements IAtomicLong {
+
+    private final RaftGroupId groupId;
+    private final String objectName;
+    private UUID objectUUID;
+
+    public AtomicLongProxy(ClientContext context, RaftGroupId groupId, String proxyName, String objectName, UUID objectUUID) {
+    ...
+    }
+}
+```
+
+On the Server side, each CP data structure (at the current moment `AtomicLong` and `AtomicRef`) should also have a UUID property:
+
+```java
+public abstract class RaftAtomicValue<T> {
+
+    private final CPGroupId groupId;
+    private final String name;
+    private UUID uuid;
+
+    public RaftAtomicValue(CPGroupId groupId, String name, UUID uuid) {
+        this.groupId = groupId;
+        this.name = name;
+        this.uuid = uuid;
+    }
+    ...
+}
+```
+
+To implement Client -> Server interaction there are two options:
+- the name-mangling scheme
+- extend Client/Server API
+
+Let's analyze each of those options
+
+#### The name-mangling scheme
+
+The name-mangling scheme has been implemented in a POC
+
+To minimize the number of changes on a client/server side, and do not recreate/update dozens of codecs, messages and Raft operations, object UUID and object name are concatenated on the Client side (the name-mangling scheme). This combined object name passes via existing API during each operation call:
+
+```java
+public class AtomicLongProxy extends ClientProxy implements IAtomicLong {
+
+    private final RaftGroupId groupId;
+    private final String objectName;
+    private UUID objectUUID;
+    ...
+    
+    public InternalCompletableFuture<Long> getAsync() {
+        ClientMessage request = AtomicLongGetCodec.encodeRequest(groupId, getCombinedObjectName());
+        ClientInvocationFuture future = new ClientInvocation(getClient(), request, name).invoke();
+        return new ClientDelegatingFuture<>(future, getSerializationService(), AtomicLongGetCodec::decodeResponse);
+    }
+
+    private String getCombinedObjectName() {
+        return objectName + "@" + objectUUID;
```

Comment:
I think this should not be `@` as it's already used to distinguish the CP group.